base/compiler/tfuncs.jl

```diff
@@ -653,6 +678,43 @@ add_tfunc(getfield, 2, 3, getfield_tfunc, 1)
 add_tfunc(setfield!, 3, 3, (@nospecialize(o), @nospecialize(f), @nospecialize(v)) -> v, 3)
 fieldtype_tfunc(@nospecialize(s0), @nospecialize(name), @nospecialize(inbounds)) =
     fieldtype_tfunc(s0, name)
+
+function fieldtype_nothrow(@nospecialize(s0), @nospecialize(name))
+    if s0 === Any || s0 === Type || DataType ⊑ s0 || UnionAll ⊑ s0
+        # We have no idea
+        return false
+    end
+
+    if !isa(name, Const) || (!isa(name.val, Symbol) || !isa(name.val, Int))
+        # Due to bounds checking, we can't say anything unless we know what
+        # the name is.
+        return false
+    end
+
+    s = instanceof_tfunc(s0)[1]
+    if isa(s, Union)
+        return _fieldtype_nothrow(s.a, name) && _fieldtype_nothrow(s.b, name)
+    end
+    return _fieldtype_nothrow(s, name)
+end
+
+function _fieldtype_nothrow(@nospecialize(s), @nospecialize(name))
+    fld = name.val
+    if isa(fld, Symbol)
+        fld = fieldindex(s, fld, false)
+    end
+    isa(fld, Int) || return false
+    ftypes = u.types
+    nf = length(ftypes)
+    if u.name === Tuple.name && fld >= nf && isvarargtype(ftypes[nf])
+        # The length of the tuple will be determined at runtime, we can't say
+        # anything
+        return false
+    end
+    (fld >= 1 && fld <= nf) || return false
```

Comment:
maybe put this condition first (such that in case `nf == 0, fld == 1, u.name == Tuple.name`, it won't try to access `ftypes[0]`)

Reply:
Sure, though I believe in that case, the tfunc would have said it's `Bottom`, so we wouldn't have reached here.